nodeup/pkg/model/kubelet.go

```diff
@@ -0,0 +1,238 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package model
+
+import (
+	"fmt"
+	"github.com/golang/glog"
+	"k8s.io/kops/nodeup/pkg/distros"
+	"k8s.io/kops/pkg/apis/kops"
+	"k8s.io/kops/pkg/flagbuilder"
+	"k8s.io/kops/pkg/kubeconfig"
+	"k8s.io/kops/pkg/systemd"
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kops/upup/pkg/fi/nodeup/nodetasks"
+)
+
+// KubeletBuilder install kubelet
+type KubeletBuilder struct {
+	*NodeupModelContext
+}
+
+var _ fi.ModelBuilder = &DockerBuilder{}
+
+func (b *KubeletBuilder) Build(c *fi.ModelBuilderContext) error {
+	kubeletConfig, err := b.buildKubeletConfig()
+	if err != nil {
+		return fmt.Errorf("error building kubelet config: %v", err)
+	}
+
+	// Add sysconfig file
+	{
+		// TODO: Dump this - just complexity!
+		flags, err := flagbuilder.BuildFlags(kubeletConfig)
+		if err != nil {
+			return fmt.Errorf("error building kubelet flags: %v", err)
+		}
+		sysconfig := "DAEMON_ARGS=\"" + flags + "\"\n"
+
+		t := &nodetasks.File{
+			Path:     "/etc/sysconfig/kubelet",
+			Contents: fi.NewStringResource(sysconfig),
+			Type:     nodetasks.FileType_File,
+		}
+		c.AddTask(t)
+	}
+
+	// Add kubelet file itself (as an asset)
+	{
+		// TODO: Extract to common function?
+		assetName := "kubelet"
+		assetPath := ""
+		asset, err := b.Assets.Find(assetName, assetPath)
+		if err != nil {
+			return fmt.Errorf("error trying to locate asset %q: %v", assetName, err)
+		}
+		if asset == nil {
+			return fmt.Errorf("unable to locate asset %q", assetName)
+		}
+
+		t := &nodetasks.File{
+			Path:     b.kubeletPath(),
+			Contents: asset,
+			Type:     nodetasks.FileType_File,
+			Mode:     s("0755"),
+		}
+		c.AddTask(t)
+	}
+
+	// Add kubeconfig
+	{
+		kubeconfig, err := b.buildKubeconfig()
+		if err != nil {
+			return err
+		}
+		t := &nodetasks.File{
+			Path:     "/var/lib/kubelet/kubeconfig",
+			Contents: fi.NewStringResource(kubeconfig),
+			Type:     nodetasks.FileType_File,
+			Mode:     s("0400"),
+		}
+		c.AddTask(t)
+	}
+
+	if b.UsesCNI {
+		t := &nodetasks.File{
+			Path: "/etc/cni/net.d/",
+			Type: nodetasks.FileType_Directory,
+		}
+		c.AddTask(t)
+	}
+
+	c.AddTask(b.buildSystemdService())
+
+	return nil
+}
+
+func (b *KubeletBuilder) kubeletPath() string {
+	kubeletCommand := "/usr/local/bin/kubelet"
```

Comment:
We have a bunch of strings here ... does it help us to make these constants? Also should we make the name of the component a const and then append the path?